packages/opentelemetry-exporter-prometheus/src/PrometheusSerializer.ts

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { MetricRecord, AggregatorKind } from '@opentelemetry/metrics';
+import { PrometheusExportType, PrometheusCheckpoint } from './types';
+import { Labels } from '@opentelemetry/api';
+import { hrTimeToMilliseconds } from '@opentelemetry/core';
+
+function escapeString(str: string) {
+  return str.replace(/\n/g, '\\n').replace(/\\(?!n)/g, '\\\\');
+}
+
+function escapeLabelValue(str: string) {
+  if (typeof str !== 'string') {
```

Comment:
this whole "escape" is wrong and doesn't respect what spec says
```
metric_name and label_name carry the usual Prometheus expression language restrictions.
label_value can be any sequence of UTF-8 characters, but the backslash (\), double-quote ("), and line feed (\n) characters have to be escaped as \\, \", and \n, respectively.
value is a float represented as required by Go's ParseFloat() function. In addition to standard numerical values, Nan, +Inf, and -Inf are valid values representing not a number, positive infinity, and negative infinity, respectively.
The timestamp is an int64 (milliseconds since epoch, i.e. 1970-01-01 00:00:00 UTC, excluding leap seconds), represented as required by Go's ParseInt() function.
```

This is the basic conversion and string builder to show what I'm talking about
```
const backslash = "\\";
const doubleQuote = '"';
const lineFeed = "\n";
const textN = "n";
const space = " ";
const arr1 = ['one', space, backslash, backslash, doubleQuote, doubleQuote, backslash, textN, backslash, textN, space, 'end', backslash, textN];
const arr2 = ['two', backslash, textN, space, backslash, backslash, doubleQuote, space, backslash, backslash, doubleQuote, space, 'end'];
function replaceSpec(arr) {
  const newArr = [];
  for (let i = 0, j = arr.length; i < j; i++) {
    if (arr[i] === backslash) {
      newArr.push(arr[i]);
      newArr.push(arr[i]);
    } else if (arr[i] === doubleQuote) {
      newArr.push(backslash);
      newArr.push(arr[i]);
    } else if (arr[i] === lineFeed) {
      newArr.push(backslash);
      newArr.push(textN);
    } else {
      newArr.push(arr[i]);
    }
  }
  return newArr;
}
const line1 = arr1.join('');
const line2 = arr2.join('');
const text = line1 + lineFeed + line2;

const shouldBe1 =
  replaceSpec(arr1).join('') +
  replaceSpec([lineFeed]).join('') +
  replaceSpec(arr2).join('');
// ensure the same test with already concatenated chars
const shouldBe2 =
  replaceSpec(line1.split('')).join('') +
  replaceSpec([lineFeed]).join('') +
  replaceSpec(line2.split('')).join('');

// convert existing text
const shouldBe3 =
  replaceSpec(text.split('')).join('');

function escapeString(str) {
  return str.replace(/\n/g, '\\n').replace(/\\(?!n)/g, '\\\\');
}
function escapeLabelValue(str) {
  if (typeof str !== 'string') {
    str = String(str);
  }
  return escapeString(str).replace(/"/g, '\\"');
}
var result = escapeLabelValue(text);
console.log('line1', line1);
console.log('line2', line2);
console.log('text', text);
console.log('result', result);

console.log('shouldBe1', shouldBe1);
console.log('shouldBe2', shouldBe2);
console.log('shouldBe3', shouldBe3);
// both conversion works fine
console.log(shouldBe2 === shouldBe1);
console.log(shouldBe3 === shouldBe1);
console.log(result === shouldBe1);
```

Reply:
Although `function escapeLabelValue(str: string)` declares the compile-time type definition, it is still possible to use OpenTelemetry without TypeScript and pure JavaScript. Thus it is possible to receive a non-string value typed label values. We have to proactively check all possible malicious typed values and convert them to expected type. Here we convert all non-string value to strings.

Also, I'm afraid I didn't get what you mean by `this whole "escape" is wrong and doesn't respect what spec says`.
As you quoted, 

> label_value can be any sequence of UTF-8 characters, but the backslash (\), double-quote ("), and line feed (\n) characters have to be escaped as \\, \", and \n, respectively.

Here we did converts backslash (\), double-quote ("), and line feed (\n) characters to the expected escaped values. 

1. for backslash: `replace(/\\(?!n)/g, '\\\\');`
2. for double quote: `escapeString(str).replace(/"/g, '\\"');`
3. and for line feed: `str.replace(/\n/g, '\\n')`

Can you elaborate on what conditions the PR did not conform to the spec?

Reply:
@legendecas I posted code example just run it. You will see that it doesn't convert correctly

Reply:
line1 = one \\""\n\n end\n
line2 = two\n \\" \\" end
concatenate them together with a line feed and run escapeLabelValue
you will get 
one \\\\\"\"\n\n end\n\ntwo\n \\\\\" \\\\\" end
which is wrong as it should be
one \\\\\"\"\\n\\n end\\n\ntwo\\n \\\\\" \\\\\" end

Reply:
Tests added as https://github.com/open-telemetry/opentelemetry-js/pull/1310/files#diff-6db9781fe72458cffe764ca6261f6926R461 . PTAL :)